tools/extra/fpgadiag/mactest.py

```diff
@@ -0,0 +1,169 @@
+            ethifs = sorted(self.ethif.items())
+            for i in ethifs:
+                print('  {:<20} {}'.format(*i))
+            print(divide)
+        else:
+            exception_quit('No ethernet interface found!')
+
+    def str2hex(self, c):
+        return int(binascii.b2a_hex(c), 16)
+
+    def read_mac_from_nvmem(self):
+        mac_number = 0
+        mac = 0
+        with open(self.args.nvmem, 'rb') as f:
+            f.seek(self.args.offset, 0)
+            for x in [40, 32, 24, 16, 8, 0]:
+                mac |= self.str2hex(f.read(1)) << x
+            mac_number = self.str2hex(f.read(1))
+        print('Read {} mac addresses from NVMEM:'.format(mac_number))
+        vendor = '{:06x}'.format(mac >> 24)
+        for i in range(mac_number):
+            mac += 1
+            mac_str = '{}{:06x}'.format(vendor, (mac & 0xffffff))
+            fmt_str = ':'.join([mac_str[x:x + 2]
+                                for x in range(0, len(mac_str), 2)])
+            self.mac.append(fmt_str)
+        for m in self.mac:
+            print('  {}'.format(m))
+        print(divide)
+
+    def compare_eth_mac_with_macrom(self):
+        result = 'PASS'
+        for m in self.mac:
+            if m not in self.ethif.values():
+                print('{} in MAC ROM, have no relative interface!'.format(m))
+                result = 'FAIL'
+        ethifs = sorted(self.ethif.items())
+        for e, m in ethifs:
+            if m not in self.mac:
+                print('{} mac {}, is not from MAC ROM'.format(e, m))
+                result = 'FAIL'
+        print('TEST {}!'.format(result))
+
+    def start(self):
+        self.get_if_and_mac_list()
+        self.read_mac_from_nvmem()
+        self.compare_eth_mac_with_macrom()
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--bus', '-B',
+                        help='Bus number of PCIe device')
+    parser.add_argument('--device', '-D',
+                        help='Device number of PCIe device')
+    parser.add_argument('--function', '-F',
+                        help='Function number of PCIe device')
+    parser.add_argument('--offset',
+                        default='0',
+                        help='read mac address from a offset address')
+    args, left = parser.parse_known_args()
+
+    args = convert_argument_str2hex(
+        args, ['bus', 'device', 'function', 'offset'])
+
+    f = FpgaFinder(args.bus, args.device, args.function)
+    devs = f.find()
+    for d in devs:
+        print(
+            'bus:0x{bus:x} device:0x{dev:x} function:0x{func:x}'.format(
+                **d))
+    if len(devs) > 1:
+        exception_quit('{} FPGAs are found\nplease choose '
+                       'one FPGA to do mactest'.format(len(devs)))
+    if not devs:
+        exception_quit('no FPGA found')
+    args.fpga_root = devs[0].get('path')
+    nvmem_path = f.find_node(devs[0].get('path'), 'nvmem', depth=7)
```

Comment:
I think the depth can be greater than 7, eg if behind a PCI bridge.
Same for 151 and 158.

Reply:
Agree that if this is traversing a /sys/class/fpga path, then it is ok.